src/waitress/adjustments.py

```diff
@@ -487,23 +525,22 @@ def parse_args(cls, argv):
 
     @classmethod
     def check_sockets(cls, sockets):
-        has_unix_socket = False
-        has_inet_socket = False
-        has_unsupported_socket = False
+        supported_families = [socket.AF_INET, socket.AF_INET6]
+        if hasattr(socket, "AF_UNIX"):
+            supported_families.append(socket.AF_UNIX)
+        if hasattr(socket, "AF_VSOCK"):
+            supported_families.append(socket.AF_VSOCK)
+
+        inet_families = (socket.AF_INET, socket.AF_INET6)
+        family = None
         for sock in sockets:
-            if (
-                sock.family == socket.AF_INET or sock.family == socket.AF_INET6
-            ) and sock.type == socket.SOCK_STREAM:
-                has_inet_socket = True
-            elif (
-                hasattr(socket, "AF_UNIX")
-                and sock.family == socket.AF_UNIX
-                and sock.type == socket.SOCK_STREAM
-            ):
-                has_unix_socket = True
-            else:
-                has_unsupported_socket = True
-        if has_unix_socket and has_inet_socket:
-            raise ValueError("Internet and UNIX sockets may not be mixed.")
-        if has_unsupported_socket:
-            raise ValueError("Only Internet or UNIX stream sockets may be used.")
+            if sock.type != socket.SOCK_STREAM or sock.family not in supported_families:
+                raise ValueError(
+                    "Only Internet, UNIX, or VSOCK stream sockets may be used."
```

Comment:
Why does this restriction exist btw? Is there an issue listening on multiple types of sockets simultaneously?

Reply:
I think it's a fair question - I don't know what the history is here but it doesn't feel like anything would fundamentally break by using multiple different socket types. Things that come to mind are any WSGI environ settings that should be dynamic that aren't right now based on the source connection.

Reply:
Each new socket type added later on will increase the number of combined checks needed. I could try removing this restriction and seeing if anything breaks.

Removing this restriction would also mean better systemd socket activation support. Systemd can pass multiple different types of sockets to apps quite easily and waitress should support that too.

Reply:
Yes. Technically there is no basis for this check. The only question is if any waitress code assumes so or depends on this.

Reply:
Part of it is that there is a different expectation with being able to receive from a UNIX socket (local, defined by file system permissions) and a TCP/IP socket. 

There is no remote peer in the case of a UNIX socket that has a port number, certain values can't be automatically provided instead they are guessed at (such as the Host header if none is provided by the remote client over a UNIX socket).

Keeping the socket types separate may not be strictly required, it's not something that has been tested or validated and adding the additional testing may not be worth it. If you want to listen on both a UNIX socket and TCP/IP then that can be two different instances of waitress (or a single front-end proxy that does support that).